File: 2017/csharp/Day1.cs
namespace AdventOfCode.CSharp.Y2017;

class Day1 : AdventOfCode.CSharp.RobotElf
{
    public Day1() : base(1) {}

    public override object Part1()
    {
        var x = Input.First();
    var s = 0;
        var q = -1;
        foreach (var c in x) {
            var j = int.Parse(c.ToString());
            if (j == q) {
                s += j;
            }
            q = j;
        }
        if (x[x.Length - 1] == x[0]) 
        {
            s += int.Parse(x[0].ToString());
        }




        return s;
    }

    public override object Part2() 
    {


        var x = Input.First();
        var s = 0;
        for (var i=0; i < x.Length; ++i)
        {
            var ni = ((x.Length / 2) + i) % x.Length;

            if (x[ni] == x[i]) 
            {
                s += int.Parse(x[i].ToString());
            }
        }


        return s;
    }
}

File: 2017/fsharp/day1.fs
namespace AdventOfCode.FSharp.Y2017

open System

module Day1 =

    let parse (input: string seq) =
        input |> Seq.map Int32.Parse |> Seq.toArray


    let part1 (input: string seq) =
        let x = input |> parse

        Array.allPairs x x
        |> Seq.filter (fun (a, b) -> a <> b && (a + b) = 2020)
        |> Seq.map (fun (a, b) -> a * b)
        |> Seq.head
        |> bigint

    let list3tuple (l: 'a list list) =
        l |> List.map (fun x -> x.[0], x.[1], x.[2])

    let part2 (input: string seq) =
        let x = input |> parse

        Array.allPairs x x
        |> Array.allPairs x
        |> Seq.filter (fun (a, (b, c)) -> a <> b && a <> c && (a + b + c) = 2020)
        |> Seq.map (fun (a, (b, c)) -> a * b * c)
        |> Seq.head
        |> bigint
